### PR TITLE
EngineLoadPrivateKey wrapper over ENGINE_load_private_key

### PR DESCRIPTION
**Pull request by merlin-northern**
Provide ability to load private key from ENGINE,
possibly HSM, SmartCard, or other.

partly inspired by Renato Aguiar https://github.com/renatoaguiar

ChangeLog:none
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

File: key.go [hunk 1]
```diff
@@ -272,6 +274,30 @@ func (key *pKey) MarshalPKIXPublicKeyDER() (der_block []byte,
...
 
+// EngineLoadPrivateKey loads a private key by id
+// the id is a pkcs#11 URI https://tools.ietf.org/html/rfc7512#section-2.3
+// Engine comes from e.g.: e,err:=openssl.EngineById("pkcs11")
+func EngineLoadPrivateKey(e *Engine, id string) (PrivateKey, error) {
+	keyID := C.CString(id)
+	defer C.free(unsafe.Pointer(keyID))
+
+	if e == nil {
+		return nil, errors.New("ENGINE_load_private_key cannot be called with NULL engine")
```
Comment (merlin-northern) [inline key.go:282]: strictly speaking it can be called, but will fail.

File: key.go [hunk 2]
```diff
@@ -272,6 +274,30 @@ func (key *pKey) MarshalPKIXPublicKeyDER() (der_block []byte,
 	return ioutil.ReadAll(asAnyBio(bio))
 }
 
+// EngineLoadPrivateKey loads a private key by id
+// the id is a pkcs#11 URI https://tools.ietf.org/html/rfc7512#section-2.3
+// Engine comes from e.g.: e,err:=openssl.EngineById("pkcs11")
+func EngineLoadPrivateKey(e *Engine, id string) (PrivateKey, error) {
+	keyID := C.CString(id)
```
Comment (oleorhagen) [inline key.go]: Suggestion:
Move after the error check

File: key.go [hunk 3]
```diff
@@ -272,6 +274,30 @@ func (key *pKey) MarshalPKIXPublicKeyDER() (der_block []byte,
...
+	if e == nil {
+		return nil, errors.New("ENGINE_load_private_key cannot be called with NULL engine")
+	}
+
+	key := C.ENGINE_load_private_key(e.e, keyID, nil, nil)
+	if key == nil {
+		return nil, errors.New("cannot load private key, ENGINE_load_private_key error")
+	}
+
+	// ref trick inspired by the work of Renato Aguiar https://github.com/renatoaguiar
```
Comment (oleorhagen) [inline key.go]: Question:
What does the trick do?

File: key.go [hunk 4]
```diff
@@ -272,6 +274,30 @@ func (key *pKey) MarshalPKIXPublicKeyDER() (der_block []byte,
...
+	p := &pKey{key: key, ref: e}
+	runtime.SetFinalizer(p, func(p *pKey) {
+		C.X_EVP_PKEY_free(p.key)
+	})
+	return p, nil
```
Comment (oleorhagen) [inline key.go:299]: Question:
Is the caller responsible for freeing this key after the function call?

Comment (oleorhagen) [reply key.go:299]: I think I see now. This is what the finalizer is for?

**Review by oleorhagen — APPROVED** [review]
I think it looks good :100:

File: key.go [hunk 5]
```diff
@@ -105,6 +106,7 @@ type PrivateKey interface {
 
 type pKey struct {
 	key *C.EVP_PKEY
+	ref interface{}
```
Comment (alfrunes) [inline key.go]: It doesn't seem clear what this `ref` field is. Could you rename/add a type or a comment?

File: key.go (hunk 3 above)
Comment (merlin-northern) [reply key.go]: guards against the freeing of the engine while the key is still in use.

Comment (merlin-northern) [reply key.go]: I updated the comment.

File: key.go (hunk 4 above)
Comment (merlin-northern) [reply key.go:299]: yes!

File: key.go (hunk 5 above)
Comment (merlin-northern) [reply key.go]: renamed to engine_ref and explained below and here reference added.

**Comment by merlin-northern** [comment]
thanks! I hope/think that I have corrected everything.

File: key.go [hunk 6]
```diff
@@ -272,6 +274,31 @@ func (key *pKey) MarshalPKIXPublicKeyDER() (der_block []byte,
...
+
+	key := C.ENGINE_load_private_key(e.e, keyID, nil, nil)
+	if key == nil {
+		return nil, errors.New("cannot load private key, ENGINE_load_private_key error")
+	}
+
+	// engine_ref trick inspired by the work of Renato Aguiar https://github.com/renatoaguiar
+	// it prevents the engine to be freed while we still use the key.
+	p := &pKey{key: key, engine_ref: e}
+	runtime.SetFinalizer(p, func(p *pKey) {
```
Comment (alfrunes) [inline key.go:296]: On a second thought; this is VERY dangerous. Looking at the other functions with `pKey` receiver, there's no call to `runtime.KeepAlive` (or `runtime.LockOSThread`) after last reference to this struct, potentially making the garbage collector routine execute the finalizer before the last cgo call returns (which might lead to undefined behavior).
Consider the following example:
```
package main

/*
#include <stdio.h>
#include <unistd.h>

void stuff(char *p) // sleep for a sec and print string
{
	sleep(1); write(1, p, 4);
}
*/
import "C"

import (
	"runtime"
	"time"
	"unsafe"
)

type Test struct {
	Foo *C.char
}

func (t *Test) DoTheStuff() {
	C.write(1, unsafe.Pointer(t.Foo), 4)
	C.stuff(t.Foo) // last reference here!
}

// A terrible example of the dangers of runtime.SetFinalizer
func main() {
	str := []byte("foo\n")
	t := &Test{Foo: (*C.char)(unsafe.Pointer(&str[0]))}
	runtime.SetFinalizer(t, func(t *Test) { str[0] = 'p' }) // imagine C.free()
	go func() {
		for {
			runtime.GC()
			time.Sleep(time.Millisecond * 500)
		}
	}()
	t.DoTheStuff()
}
```
When executed, the finalizer will run while the cgo call is in progress.

**Review by alfrunes — CHANGES_REQUESTED** [review]
Make sure all `cgo` calls are protected with `runtime.KeepAlive` (or `runtime.LockOSThread`)

File: key.go (hunk 6 above)
Comment (merlin-northern) [reply key.go:296]: discussed on slack. thanks @alfrunes

**Review by alfrunes — APPROVED** [review]
As this PR does not introduce any harm, I'll approve this one. However, we should follow up on the above [issue](https://tracker.mender.io/browse/MEN-3877) ASAP.

**Review by oleorhagen — APPROVED** [review]

**Comment by merlin-northern** [comment]
> As this PR does not introduce any harm, I'll approve this one. However, we should follow up on the above [issue](https://tracker.mender.io/browse/MEN-3877) ASAP.

I am on it!

**Pull request merged**